Expose data on api

baltazarix opened the pull request:
expose data from models `Case`, `Story` at `/data.json`

closes #21

baltazarix committed:
implemented to_JSON method in Case, Story models

baltazarix committed:
added route (/data.json) which returns data in json format

cuducos commented on victims/models.py:
```diff
@@ -27,6 +29,20 @@ def is_valid(self):
 
         return True
 
+    def to_JSON(self):
```
`to_JSON` is a syntax that [does not follow PEP8](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables):

> **Method Names and Instance Variables**
> 
> Use the function naming rules: lowercase with words separated by underscores as necessary to improve readability.

Can we rename it to `to_json`?

cuducos commented on victims/models.py:
```diff
@@ -27,6 +29,20 @@ def is_valid(self):
+    def to_JSON(self):
+        """
+        Serialize object to json
+        """
```
This short docstring could be in one:

```
def to_json(self):
    """Serialize object to json"""
```

But is this docstring really necessary? I feel like the name `to_json` says it all and the docstring is redundant. What do you think about it?

cuducos commented on victims/models.py:
```diff
@@ -27,6 +29,20 @@ def is_valid(self):
+        """
+        Serialize object to json
+        """
+        d = {
```
What does `d` stands for? May we have a more meaningful variable name here?

cuducos commented on victims/models.py:
```diff
@@ -48,3 +64,18 @@ def tags_and_colors(self):
+            "tags": self.tags,
+            "stories": [story.to_JSON() for story in self.stories],
+        }
+        return json.dumps(d, sort_keys=True)
```
Just the same as my comments in the other `dataclass` ; )

baltazarix committed:
small improvements...
- corrected method names according to PEP8
- removed redundant docstrings
- more meaningful variable name inside 'to_json' method

baltazarix commented:
@cuducos, Just pushed changes according to your comments. Thanks for time you've spent on this review and also for right remarks.

cuducos commented on victims/models.py:
```diff
@@ -27,6 +29,17 @@ def is_valid(self):
+            "image_or_video": self.image_or_video,
+            "summary": self.summary,
+            "case_id": self.case_id,
+        }
```
Any special reason to not use `dict(self)` and keep the descriptive verbose approach? IMHO this is technical debt.

baltazarix replied:
I used variable here cause I had to somehow serialize `datetime.date`. After my research on network I can propose to implement custom json encoder class, which will handle `datetime.date` value. With usage of `dataclasses.asdict`. Then whole `to_json` method may look like this:

```python
    def to_json(self):
        return json.dumps(asdict(self), cls=DateEncoder, sort_keys=True)
```

cuducos replied:
This is a good way. Alternatively simpler though:

```python
def to_json(self):
    data = asdict(self)
    data['when'] = date['when'].strftime("%Y-%m-%d")
    return json.dumps(data, sort_keys=True)
```

cuducos commented on victims/models.py:
```diff
@@ -48,3 +61,15 @@ def tags_and_colors(self):
+            "city": self.city,
+            "tags": self.tags,
+            "stories": [story.to_JSON() for story in self.stories],
+        }
```
Any special reason to not use `dict(self)` and keep the descriptive verbose approach? IMHO this is technical debt.

cuducos commented on victims/__init__.py:
```diff
@@ -47,3 +47,10 @@ def force_https(request):
+@app.route("/data.json")
+async def data(request):
+    cases = await get_cases()
+    cases = [case.to_JSON() for case in cases]
```
There's no `case.to_JSON()` anymore, right?

baltazarix replied:
you're right, my bad

cuducos commented:
> Just pushed changes according to your comments.

I think you missed some points. Could you review or come back with questions to help me clarify my points? So sorry if I was unable to make myself clearer : (

turicas commented on victims/models.py:
```diff
@@ -48,3 +61,15 @@ def tags_and_colors(self):
+            "state": self.state,
+            "city": self.city,
+            "tags": self.tags,
+            "stories": [story.to_JSON() for story in self.stories],
```
This line should not encode `story` as JSON since the whole `items` is encoded in the end.

baltazarix replied:
good point, thanks

turicas replied:
@baltazarix do you mind adding a simple test for this endpoint? [It's pretty simple](https://sanic.readthedocs.io/en/latest/sanic/testing.html) and can avoid assure we're serializing data correctly.

baltazarix replied:
@turicas, sure! I was thinking about it but I've had no idea how test scenario should look like. Only ideas I have are:
- checking for status code
- create some test data (e.g. 1 `Case` instance, 2 `Story` instances), then compare response content.

Feel free to tell me if I'm missing something :)

baltazarix replied:
> Take a look at existing `test_home_contents`: we test some contents that we expect to see in the response. Name of the stories, maybe dates formatted as str (as they are the bottleneck of this serialization) etc…

Yep, I wanted to make something similar, sorry if I wasn't clear enough.
